File: app/src/main/java/com/toandoan/todoapp/screen/main/MainAdapter.java
package com.toandoan.todoapp.screen.main;

import android.arch.paging.PagedListAdapter;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.DiffCallback;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.toandoan.todoapp.BR;
import com.toandoan.todoapp.R;
import com.toandoan.todoapp.data.model.Repo;
import com.toandoan.todoapp.databinding.ItemRepositoryBinding;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by doan.van.toan on 12/26/17.
 */

public class MainAdapter extends RecyclerView.Adapter<MainAdapter.ViewHolder> {
    private List<Repo> mRepos;
    private LayoutInflater mInflater;
    private RepoItemClickListener mRepoItemClickListener;

    protected MainAdapter() {
        mRepos = new ArrayList<>();
    }

    public void setRepoItemClickListener(RepoItemClickListener repoItemClickListener) {
        mRepoItemClickListener = repoItemClickListener;
    }

    public void addData(List<Repo> repos) {
        if (repos == null) {
            return;
        }
        mRepos.addAll(repos);
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (mInflater == null) {
            mInflater = LayoutInflater.from(parent.getContext());
        }
        ItemRepositoryBinding binding = DataBindingUtil.inflate(mInflater,
                R.layout.item_repository, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bindData(mRepos.get(position));
    }

    @Override
    public int getItemCount() {
        return mRepos != null ? mRepos.size() : 0;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private ItemRepositoryBinding mBinding;

        public ViewHolder(ItemRepositoryBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
        }

        public void bindData(Repo repo) {
            mBinding.setRepo(repo);
            mBinding.setPosition(String.valueOf(getAdapterPosition() + 1));
            mBinding.setListener(mRepoItemClickListener);
            mBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/toandoan/todo/screen/main/MainAdapter.kt
package com.toandoan.todo.screen.main

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.TextView
import com.toandoan.todo.R
import com.toandoan.todo.data.model.Task

class MainAdapter(var tasks: MutableList<Task>,
    var listenner: OnItemCheckListenner) : RecyclerView.Adapter<MainAdapter.ViewHolder>() {


  fun updateData(task: MutableList<Task>) {
    this.tasks.addAll(task)
    notifyDataSetChanged()
  }

  fun updateData(task: Task) {
    this.tasks.add(0, task)
    notifyItemInserted(0)
  }

  fun clearData() {
    this.tasks.clear()
    notifyDataSetChanged()
  }

  override fun getItemCount(): Int {
    return tasks.size
  }

  override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
    var view: View = LayoutInflater.from(parent?.context)
        .inflate(R.layout.item_task, parent, false)
    return ViewHolder(view)
  }

  override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
    holder?.bindData(tasks[position], listenner)
  }

  class ViewHolder : RecyclerView.ViewHolder {
    var textTitle: TextView? = null
    var textDescription: TextView? = null
    var checkActive: CheckBox? = null

    constructor(itemView: View?) : super(itemView) {
      textTitle = itemView?.findViewById(R.id.textTitle)
      textDescription = itemView?.findViewById(R.id.textDescription)
      checkActive = itemView?.findViewById(R.id.checkedActive)

    }

    fun bindData(task: Task, listenner: OnItemCheckListenner) {
      textTitle?.text = task.title
      if (task?.description != null) {
        textDescription?.text = task.description
      } else {
        textDescription?.visibility = View.GONE
      }
      checkActive?.isChecked = task.isActive!!
      checkActive?.setOnCheckedChangeListener(
          CompoundButton.OnCheckedChangeListener { compoundButton: CompoundButton,
              isChecked: Boolean ->
            task.isActive = isChecked
            listenner?.onCheckChange(task)
          })
    }
  }
}